File: Sorts/python/quick_sort.py
def quick_sort(arr, low, high):
    print("low: ", low)
    print("high: ", high)

    if low < high:
        j = partition(arr, low, high)
        quick_sort(arr, low, j)
        quick_sort(arr, j+1, high)


def partition(arr, low, high):
    i = low
    j = high
    pivot = arr[low]

    while i < j:
        while i <= high and arr[i] <= pivot:
            i += 1
        
        while j >= low and arr[j] > pivot:
            j -= 1
        
        if i < j:
            swap(arr, i, j)
    swap(arr, low, j)
    return j
    
def swap(arr, indexOne, indexTwo):
    arr[indexOne], arr[indexTwo] = arr[indexTwo], arr[indexOne]

list = [ 19, 15, 11, 13, 1, 6, 14, 2 ]
print("before: ", list)
quick_sort(list, 0, len(list) - 1)
print("after: ", list)

File: Sorts/js/quick_sort.js
function quick_sort(arr, low, high) {
  if (low < high) {
    let j = partition(arr, low, high);
    quick_sort(arr, low, j);
    quick_sort(arr, j + 1, high);
  }
}

function partition(arr, low, high) {
  let i = low;
  let j = high;
  let pivot = arr[low];

  while (i < j) {
    while (i >= high && arr[i] <= pivot) {
      i++;
    }

    while (arr[j] > pivot) {
      j--;
    }

    if (i < j) {
      swap(arr, i, j);
    }
  }

  swap(arr, low, j);
  return j;
}

function swap(arr, indexOne, indexTwo) {
  let temp = arr[indexOne];
  arr[indexOne] = arr[indexTwo];
  arr[indexTwo] = temp;
}

arr = [19, 15, 11, 13, 1, 6, 14, 2];
console.log("before: ", arr);
quick_sort(arr, 0, arr.length - 1);
console.log("After: ", arr);
